lib/cylc/profiling/git.py

```diff
@@ -0,0 +1,100 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Provides python wrappers to certain git commands."""
+
+import os
+from subprocess import (Popen, PIPE, CalledProcessError, check_call)
+
+
+class GitCheckoutError(Exception):
+    """Exception to be raised if a git checkout command fails."""
+    pass
+
+
+def describe(ref=None):
+    """Returns stdout of the `git describe <COMMIT>` command."""
+    try:
+        cmd = ['git', 'describe']
+        if ref:
+            cmd.append(ref)
+        return Popen(cmd, stdout=PIPE).communicate()[0].strip()
+    except CalledProcessError:
+        return None
+
+
+def is_ancestor_commit(commit1, commit2):
+    """Returns True if commit1 is an ancestor of commit2."""
+    try:
+        ancestor = Popen(['git', 'merge-base', commit1, commit2],
+                         stdout=PIPE).communicate()[0].strip()
+        return ancestor == commit1
+    except CalledProcessError:
+        return False
+
+
+def checkout(branch, delete_pyc=False):
+    """Checkouts the git branch with the provided name."""
+    try:
+        cmd = ['git', 'checkout', '-q', branch]
+        print '$ ' + ' '.join(cmd)
+        check_call(cmd)
+    except CalledProcessError:
+        raise GitCheckoutError()
+    try:
+        if delete_pyc:
+            cmd = ['find', 'lib', '-name', r'\*.pyc', '-delete']
+            print '$ ' + ' '.join(cmd)
+            check_call(cmd, stdout=open(os.devnull, 'wb'))
+    except CalledProcessError:
+        pass
+
+
+def get_commit_date(commit):
+    """Returns the commit date (in unix time) of the profided commit."""
+    return int(Popen(['git', 'show', '-s', '--format=%at', commit],
+                     stdout=PIPE).communicate()[0].split()[-1])
+
+
+def order_versions_by_date(versions):
+    """Orders a list of version objects by the date of the most recent
+    commit."""
+    versions.sort(key=lambda x: get_commit_date(x['id']))
+
+
+def order_identifiers_by_date(versions):
+    """Orders a list of git identifiers by the date of the most recent
+    commit."""
+    versions.sort(key=lambda x: get_commit_date(x))
+
+
+def has_changes_to_be_committed():
+    """Returns True if there are any un-committed changes to the working
+    copy."""
+    git_status = Popen(['git', 'status'], stdout=PIPE).communicate()[0]
+    if 'Changes to be committed' in git_status:
+        return True
+    if 'Changed but not updated' in git_status:
+        return True
+    return False
+
+
+def is_git_repo():
+    """Returns true if we are currently within a git repository."""
+    proc = Popen('[ -d .git ] || git rev-parse --git-dir > /dev/null 2>&1',
```

Comment:
Why not just run `git rev-parse` (i.e. no need to test for presence of `.git/`) and let `git` works it out?